### PR TITLE
[WIP][PoC][testing] Python array.array interface

### PR DESCRIPTION
**Pull request by mbargull**
\[ DISCLAIMER: This is no production-ready code, just for testing and unfinished, plus I know next to nothing about how to efficiently work with Python C-API :wink: \]

This is in reference to
- https://github.com/msoos/cryptominisat/issues/513, and
- https://github.com/conda/conda/pull/7676

I put together some non-polished interface to be able to pass clauses to `Solver.add_clauses` as a Python `array.array("i")` object. This is so we can compare the time taken on processing the clauses as a simple stream of integers (separated by zeros) to processing them as iterables of iterables, e.g., list of tuples, which the current interface expects.
Since the time taken to scan an integer array is practically zero (if we need to see all values anyways, that is), the remaining time of adding clauses can be assumed to be spent outside of the Python interface, i.e., be admeasured to the core library.

I'll follow up with some measurement results..

### CODE REVIEW

**Comment by jakirkham** [comment]
So I haven't dug into this code at all, which means my comments could be totally irrelevant and/or addressed. Just came across it while reading issue ( https://github.com/msoos/cryptominisat/issues/513 ).

Not sure if you are aware (or how much it matters for this use case), but Python has a [buffer protocol]( https://docs.python.org/3/c-api/buffer.html ) (supported in both Python 2 and Python 3), which allows zero copy sharing of data between different Python representations  (e.g. `array.array`, `bytes`, `numpy.ndarray`, etc.). This can also be used to wrap `std::vector` or other C/C++ array objects and have them representable in Python cheaply. The buffer protocol can be implemented in [C/C++]( https://jakevdp.github.io/blog/2014/05/05/introduction-to-the-python-buffer-protocol/ ). Also can leverage utilities in [Cython]( https://cython.readthedocs.io/en/latest/src/userguide/buffer.html ), [pybind11]( https://pybind11.readthedocs.io/en/stable/advanced/pycpp/numpy.html#buffer-protocol ), [Boost]( https://stackoverflow.com/a/17759608 ), etc. Happy to answer questions if it helps.

**Comment by mbargull** [comment]
The following is in reference to  and shows `conda`-specific results:

----
*Before* merging with the recent improvements on [`master`](https://github.com/msoos/cryptominisat/issues/513#issuecomment-415556691) and with `5.6.4` as a base we get the following results:

```sh
$ export CONDA_CHANNELS=defaults CONDA_INSTRUMENTATION_ENABLED=1 CONDA_TIMED_LOGGING=1

$ # 5.6.4 + python-array + ClauseArray
$ conda create -nx --dry-run -v anaconda
Solving environment: ...working... === time_recorder total time and calls ===
TOTAL   116.266       853 conda.common.logic.Clauses.LinearBound_
TOTAL   104.667        62 conda.common.logic.Clauses._run_sat
TOTAL   217.304        12 conda.common.logic.Clauses.minimize
TOTAL   104.670        62 conda.common.logic.Clauses.sat
TOTAL     7.272        62 conda.common.logic.CryptoMiniSatSolver.invoke
TOTAL     1.641        62 conda.common.logic.CryptoMiniSatSolver.process_solution
TOTAL    62.973        62 conda.common.logic.CryptoMiniSatSolver.setup
TOTAL    24.026        62 conda.common.logic.PycoSatSolver.setup
TOTAL    75.510        62 conda.common.logic.SatSolver.run
TOTAL    20.790         2 conda.resolve.Resolve.gen_clauses
TOTAL    13.895         2 conda.resolve.Resolve.get_reduced_index
TOTAL   232.203         1 conda.resolve.Resolve.solve

$ # 5.6.4 + python-array + ClauseList
$ conda create -nx --dry-run -v anaconda
Solving environment: ...working... === time_recorder total time and calls ===
TOTAL    79.806       853 conda.common.logic.Clauses.LinearBound_
TOTAL   176.648        62 conda.common.logic.Clauses._run_sat
TOTAL   253.282        12 conda.common.logic.Clauses.minimize
TOTAL   176.652        62 conda.common.logic.Clauses.sat
TOTAL     7.336        62 conda.common.logic.CryptoMiniSatSolver.invoke
TOTAL     1.652        62 conda.common.logic.CryptoMiniSatSolver.process_solution
TOTAL   119.002        62 conda.common.logic.CryptoMiniSatSolver.setup
TOTAL    39.833        62 conda.common.logic.PycoSatSolver.setup
TOTAL   131.348        62 conda.common.logic.SatSolver.run
TOTAL    20.426         2 conda.resolve.Resolve.gen_clauses
TOTAL    14.192         2 conda.resolve.Resolve.get_reduced_index
TOTAL   270.933         1 conda.resolve.Resolve.solve
```
Here, the most important parts are
```
TOTAL    62.973        62 conda.common.logic.CryptoMiniSatSolver.setup
TOTAL    24.026        62 conda.common.logic.PycoSatSolver.setup
```
and
```
TOTAL   119.002        62 conda.common.logic.CryptoMiniSatSolver.setup
TOTAL    39.833        62 conda.common.logic.PycoSatSolver.setup
```
. They show the total time taken for `62` invocations of `pycryptosat.Solver.add_clauses` with an `array.array("i")` or a `list` of `tuple`s of `int` as the input type for the set of clauses.
We see that
1. In this case the extra time taken to process the tuple list amounts to about half of the overall `add_clauses` runtime, and
2. The time taken by `pycosat`/PicoSAT is considerably slower.

----
*After* merging the latest improvements up to commit 10517a110f900ecea721b845fb76245be0b9fe6f, we get the following:
```
$ # after merging master (10517a1) + python-array + ClauseArray
$ conda create -nx --dry-run -v anaconda
Solving environment: ...working... === time_recorder total time and calls ===
TOTAL   112.682       853 conda.common.logic.Clauses.LinearBound_
TOTAL    99.651        62 conda.common.logic.Clauses._run_sat
TOTAL   208.951        12 conda.common.logic.Clauses.minimize
TOTAL    99.655        62 conda.common.logic.Clauses.sat
TOTAL     7.293        62 conda.common.logic.CryptoMiniSatSolver.invoke
TOTAL     1.587        62 conda.common.logic.CryptoMiniSatSolver.process_solution
TOTAL    58.094        62 conda.common.logic.CryptoMiniSatSolver.setup
TOTAL    23.862        62 conda.common.logic.PycoSatSolver.setup
TOTAL    70.517        62 conda.common.logic.SatSolver.run
TOTAL    20.656         2 conda.resolve.Resolve.gen_clauses
TOTAL    14.222         2 conda.resolve.Resolve.get_reduced_index
TOTAL   223.477         1 conda.resolve.Resolve.solve

$ # after merging master (10517a1) + python-array + ClauseList
$ conda create -nx --dry-run -v anaconda
Solving environment: ...working... === time_recorder total time and calls ===
TOTAL    81.224       853 conda.common.logic.Clauses.LinearBound_
TOTAL   150.507        62 conda.common.logic.Clauses._run_sat
TOTAL   229.594        12 conda.common.logic.Clauses.minimize
TOTAL   150.511        62 conda.common.logic.Clauses.sat
TOTAL     7.391        62 conda.common.logic.CryptoMiniSatSolver.invoke
TOTAL     1.659        62 conda.common.logic.CryptoMiniSatSolver.process_solution
TOTAL    92.382        62 conda.common.logic.CryptoMiniSatSolver.setup
TOTAL    40.174        62 conda.common.logic.PycoSatSolver.setup
TOTAL   104.823        62 conda.common.logic.SatSolver.run
TOTAL    19.895         2 conda.resolve.Resolve.gen_clauses
TOTAL    13.768         2 conda.resolve.Resolve.get_reduced_index
TOTAL   245.838         1 conda.resolve.Resolve.solve
```
, i.e.,
```
TOTAL    58.094        62 conda.common.logic.CryptoMiniSatSolver.setup
TOTAL    23.862        62 conda.common.logic.PycoSatSolver.setup
```
and
```
TOTAL    92.382        62 conda.common.logic.CryptoMiniSatSolver.setup
TOTAL    40.174        62 conda.common.logic.PycoSatSolver.setup
```
which shows a nearly 10 % improvement in the `array.array("i")` case (i.e., improvements to the core lib) and nearly 25 % improvements for the `List[Tuple[Int]]` case, i.e., also including improvements to the Python interface.

----
If we take the `pycosat` numbers as a benchmark, there is still some room for improvement -- but getting a 10 % time reduction already is nice!

**Comment by mbargull** [comment]
@jakirkham, yes using the buffer protocol would make most sense. I didn't use it because 1) I just hacked something together to get numbers and didn't invest too much time learning the Python C-API and 2) I saw
> An array can only expose its contents via the old-style buffer interface. This limitation does not apply to Python 3, where `memoryview` objects can be constructed from arrays, too. Array elements may be multi-byte values.

in https://docs.python.org/2/c-api/buffer.html (stupid Python 2 support as always...) which made me not look into it further on how to workaround that. I used `array.array` because I'm pretty sure pulling in `numpy` (and thus BLAS... fat `defaults::mkl`...) is a no go for `conda` due to size considerations. (But I'd soo much love to use `numba` :sob:)

----
Either way, I'm also not entirely sure if we'll use arrays/buffers anyway, because if we can leverage what @msoos explained in https://github.com/msoos/cryptominisat/issues/513#issuecomment-415503020 (and https://github.com/msoos/cryptominisat/issues/513#issuecomment-415505531), I'd be more than happy to not be able to store the clauses at all on the Python side but let CMS take care of it :smile:.

**Comment by mbargull** [comment]
> I'm also not entirely sure

This is also the reason this branch is not polished up and misses some `Py_DECREF`s and stuff.

**Comment by mbargull** [comment]
To create some simpler cases/benchmarks, we can also just mindlessly create a lot of clauses and variables with the following script:
```python
from array import array
from collections import defaultdict
from itertools import chain
from time import time

import pycryptosat
def cms_setup(clauses, m):
    solver = pycryptosat.Solver()
    solver.add_clauses(clauses)
    return solver
def cms_solve(solver):
    return solver.solve()

import pycosat
def ps_setup(clauses, m):
    return pycosat.itersolve(clauses, vars=m)
def ps_solve(iter_sol):
    return next(iter_sol)

n = 500000
clause_list = list(chain.from_iterable(
    [(-x, x + n), (-x, x + n)] for x in range(1, n+1)
))
clause_array = array("i", chain.from_iterable(c + (0,) for c in clause_list)) 
m = max(clause_array)


all_times = defaultdict(list)
for solver_type, setup, solve in (("CMS", cms_setup, cms_solve), ("Pycosat", ps_setup, ps_solve)):
    for clauses_type, clauses in (("list", clause_list), ("array", clause_array)):
        for _ in range(5):
            t = time()
            solver = setup(clauses, m)
            all_times["setup {:<7} {:<5}".format(solver_type, clauses_type)].append(time() - t)
            t = time()
            solve(solver)
            all_times["solve {:<7} {:<5}".format(solver_type, clauses_type)].append(time() - t)
for text, times in sorted(all_times.items()):
    print(text, *map("{:5.3f}".format, sorted(times)))
```

which gives us
```
# 5.6.4 + python-array
setup CMS     array 0.334 0.334 0.339 0.340 0.352
setup CMS     list  0.571 0.581 0.597 0.606 0.611
setup Pycosat array 0.140 0.141 0.142 0.142 0.143
setup Pycosat list  0.207 0.216 0.217 0.218 0.224
solve CMS     array 0.094 0.094 0.095 0.095 0.099
solve CMS     list  0.083 0.090 0.094 0.095 0.096
solve Pycosat array 1.055 1.060 1.064 1.064 1.068
solve Pycosat list  1.052 1.053 1.054 1.063 1.076
```
and
```
# after merging master (10517a1) + python-array
setup CMS     array 0.314 0.319 0.320 0.320 0.325
setup CMS     list  0.503 0.514 0.526 0.536 0.545
setup Pycosat array 0.140 0.141 0.142 0.142 0.145
setup Pycosat list  0.213 0.213 0.213 0.213 0.214
solve CMS     array 0.093 0.093 0.094 0.094 0.095
solve CMS     list  0.083 0.092 0.094 0.095 0.103
solve Pycosat array 1.051 1.061 1.061 1.065 1.065
solve Pycosat list  1.054 1.054 1.070 1.073 1.102
```
, which should be far more easy to work with than running those `conda create ... anaconda` cases :+1:.

----
One think to note here, is that we tell `pycosat` beforehand how many variables there are through `vars=m`. This has a noticeable performance impact as can be seen if we run the same script but with `vars=m` removed:
```
setup CMS     array 0.310 0.312 0.312 0.313 0.322
setup CMS     list  0.489 0.505 0.519 0.534 0.535
setup Pycosat array 0.159 0.171 0.174 0.174 0.175
setup Pycosat list  0.217 0.230 0.230 0.230 0.232
solve CMS     array 0.094 0.094 0.095 0.096 0.096
solve CMS     list  0.086 0.091 0.093 0.096 0.096
solve Pycosat array 1.042 1.047 1.054 1.055 1.060
solve Pycosat list  1.043 1.045 1.046 1.046 1.069
```
, i.e., we get `174 ms` instead of `142 ms` with `vars=m` for the `array.array("i")` case and `230 ms` instead of `213 ms` in the `List[Tuple[Int]]` case.
I just noticed that now and didn't yet look at how CMS behaves if we run
```cpp
self->cmsat->new_vars(max_var-(long int)self->cmsat->nVars()+1);
```
with the appropriate `max_var` beforehand..

**Comment by mbargull** [comment]
Oh, and I forgot to mention that this kind of interface addition could seem like a very special purpose thing which might not be incredibly useful in the general case. Meaning, you'd have to weigh if it makes sense to create and maintain such an interface.

But this is no show-stopper for using something like this in `conda`, since there is always the possibility for the Conda team to create a custom Python interface like @mcg1969 mentioned in https://github.com/msoos/cryptominisat/issues/513#issuecomment-415226906:
> If the compiled layer here is not the bottleneck, and only the Python interface, then the likelihood is high that we can develop a faster interface for our purposes. Maybe we do implement some sort of "clause queue" on the C++ side that allows us to incrementally load and roll back the clause set on the C++ side, as needed, even if we can't exploit an actual incremental solve.

File: python/src/pycryptosat.cpp [hunk 1]
```diff
@@ -336,6 +336,100 @@ static PyObject* add_clause(Solver *self, PyObject *args, PyObject *kwds)
...
+        return 0;
+    }
+    const char *typecode_cstr = PyBytes_AsString(typecode_bytes);
+    if (typecode_cstr == NULL) {
+        PyErr_SetString(PyExc_ValueError, "invalid clause array: typecode string is NULL");
+        return 0;
+    }
+    PyObject *py_array_address = PyTuple_GetItem(buffer_info, 0);
+    long array_length = PyLong_AsLong(PyTuple_GetItem(buffer_info, 1));
+    if (array_length < 9) {
```
Comment (mbargull) [inline python/src/pycryptosat.cpp]:
```diff
-    if (array_length < 9) {
+    if (array_length < 0) {
```
:no_mouth: https://www.urbandictionary.com/define.php?term=Fat%20Finger%20Disease

**Comment by mbargull** [comment]
@msoos, am I right to assume that `Solver::back_number_from_outside_to_outer_tmp` is a scratch pad so it's fine to avoid copying it in `Solver::addClause` and it's fine that it's sorted inplace?
I'm really just trying to pick the lowest hanging fruit there are -- but don't know the code/context at all, so I kind of feel like a bull in a china shop and don't know for certain what I'm doing here :sweat_smile: 

----
Avoiding that copy would give us another `10-15 %` or so time reduction..
```
$ # after merging master (10517a1) + max_var + addClauseInt + ClauseArray
$ conda create -nx --dry-run -v anaconda
Solving environment: ...working... === time_recorder total time and calls ===
TOTAL   114.100       853 conda.common.logic.Clauses.LinearBound_
TOTAL    82.293        62 conda.common.logic.Clauses._run_sat
TOTAL   193.810        12 conda.common.logic.Clauses.minimize
TOTAL    82.296        62 conda.common.logic.Clauses.sat
TOTAL     7.321        62 conda.common.logic.CryptoMiniSatSolver.invoke
TOTAL     1.625        62 conda.common.logic.CryptoMiniSatSolver.process_solution
TOTAL    41.596        62 conda.common.logic.CryptoMiniSatSolver.setup
TOTAL    23.225        62 conda.common.logic.PycoSatSolver.setup
TOTAL    54.400        62 conda.common.logic.SatSolver.run
TOTAL    20.762         2 conda.resolve.Resolve.gen_clauses
TOTAL    14.158         2 conda.resolve.Resolve.get_reduced_index
TOTAL   207.654         1 conda.resolve.Resolve.solve

$ # after merging master (10517a1) + max_var + addClauseInt + ClauseList
$ conda create -nx --dry-run -v anaconda
Solving environment: ...working... === time_recorder total time and calls ===
TOTAL    80.027       853 conda.common.logic.Clauses.LinearBound_
TOTAL   124.768        62 conda.common.logic.Clauses._run_sat
TOTAL   203.609        12 conda.common.logic.Clauses.minimize
TOTAL   124.772        62 conda.common.logic.Clauses.sat
TOTAL     7.302        62 conda.common.logic.CryptoMiniSatSolver.invoke
TOTAL     1.651        62 conda.common.logic.CryptoMiniSatSolver.process_solution
TOTAL    69.651        62 conda.common.logic.CryptoMiniSatSolver.setup
TOTAL    37.726        62 conda.common.logic.PycoSatSolver.setup
TOTAL    82.433        62 conda.common.logic.SatSolver.run
TOTAL    20.319         2 conda.resolve.Resolve.gen_clauses
TOTAL    14.003         2 conda.resolve.Resolve.get_reduced_index
TOTAL   219.326         1 conda.resolve.Resolve.solve
```
```
$ # after merging master (10517a1) + max_var + addClauseInt + ClauseArray
TOTAL    41.596        62 conda.common.logic.CryptoMiniSatSolver.setup
TOTAL    23.225        62 conda.common.logic.PycoSatSolver.setup
```
```
$ # after merging master (10517a1) + max_var + addClauseInt + ClauseList
TOTAL    69.651        62 conda.common.logic.CryptoMiniSatSolver.setup
TOTAL    37.726        62 conda.common.logic.PycoSatSolver.setup
```
```
$ # after merging master (10517a1) + max_var + addClauseInt
$ python -c '...'  # script from https://github.com/msoos/cryptominisat/pull/515#issuecomment-415791578
setup CMS     array 0.212 0.213 0.214 0.215 0.222
setup CMS     list  0.348 0.371 0.383 0.383 0.386
setup Pycosat array 0.139 0.139 0.140 0.141 0.143
setup Pycosat list  0.208 0.214 0.214 0.214 0.219
solve CMS     array 0.084 0.084 0.085 0.086 0.088
solve CMS     list  0.084 0.086 0.093 0.095 0.097
solve Pycosat array 1.036 1.047 1.048 1.055 1.069
solve Pycosat list  1.041 1.047 1.053 1.054 1.062
```

**Comment by msoos** [comment]
Hi, this looks cool, thank you so much!

The only think I ask is that you don't add the `addClauseInt` -- it's not really used, and it's not necessary. Can you please add a commit that removes all changes to `solver.cpp` and `solver.h`? It's confusing me -- there is already too many internal clause addition functions (with various names... sorry...) and this one is not used for anything other than calling the real function in a way that it was called before, as far as I could understand. I guess it was part of an older change set that made it in? If it actually does anything important, please let me know -- my cursory check seems not to find a reason to have it :) Thanks!

**Comment by msoos** [comment]
Regarding `max_var` -- yes, setting that right makes a big difference. I (of you, if you like) can also expose that as a python API, then you can set it right first and the solver doesn't need to call `add_new_vars` all the time :) It can save quite a bit. Should we give that a shot or it's OK now already?

**Comment by mbargull** [comment]
Hi @msoos and thanks for taking a look at this.

Sorry I made this more confusing than it should be, I got carried away with this :wink:.
So, to explain, this pull request is not to really request to pull this in, just to have a place to try things out and discuss. It currently incorporates 3 changes:
1. Adding support for passing over the clauses as an integer array.
   To me this one can be a bit controversial since it seems like a very special purpose thing targeted at a possible implementation in `conda`. Plus it probably needs some additional work.
2. Adding `max_var` to `add_clauses`.
   The one is a non-controversial small addition that makes sense in any case :smile:.
3. Avoiding to copy `back_number_from_outside_to_outer_tmp` when adding a clause via `Solver::add_clause_outer`.
   Unless there are reasons not to modify `back_number_from_outside_to_outer_tmp` in-place, this change also makes sense because it offers a noticeable performance improvement, i.e.,
   - `41.596` vs `48.374` seconds (`-14 %`) spent in `add_clauses` for `conda` with `ClauseArray`,
   - `69.651` vs `84.107` seconds (`-17 %`) spent in `add_clauses` for `conda` with `ClauseList`,
   - `0.214` vs `0.240` seconds (`-11 %`) spent in `add_clauses` for that dummy Python script.

   The name `addClauseInt` is bad and just that way because I didn't think of anything better. If you have a better suggestion how to avoid the copying of `back_number_from_outside_to_outer_tmp` at https://github.com/msoos/cryptominisat/blob/879c4ac0c9fae80a767e5c8f61797a6965889d5a/src/solver.cpp#L653 that looks nicer and doesn't have to introduce an additional private method, I'll gladly change it.

So overall, I wouldn't want this PR to be merged at all, but rather open separate PRs (with polished code) for those part that you deems worthy to be merged.

**Comment by mbargull** [comment]
Oh, regarding that `addClauseInt`, I missed that part of your sentence before:
> and this one is not used for anything other than calling the real function in a way that it was called before, as far as I could understand. 

So, yeah, it's not exactly called like before but with a subtle change, i.e., with `vector<Lit>&`  instead of `const vector<Lit>&`, so that we can change that parameter in-place. And the only time we make use of that change is when we call it with the temporary storage `back_number_from_outside_to_outer_tmp` in `add_clause_outer`. This is just so we avoid a copy there, since I assumed it is fine that we change the temporary storage in-place -- I'm not 100 % sure that assumption is correct, but I think it is ^^.

**Comment by msoos** [comment]
(First a general note: you people are crazy good at this, where the hell did you learn to do this so professionally? I thought I was pretty OK, but damn)

For 1: I don't mind it actually. We can think about it later, but it seems OK.
For 2: I agree, it looks perfect!
For 3: wow, that's a nice catch! The problem is that `addClause` can recursively call itself (yeah... nasty) so I was incredibly conservative around stuff that touches `addClause`. However, you managed to catch a nice optimisation! Cool, thanks a ton!

**Comment by msoos** [comment]
Thanks, I have now pushed a few commits (I made them related to #513 so you can track). It should copy less now. The performance difference for the startup remains vs pycosat but it's reduced. Thanks for the script by the way, cool stuff, it helped -- though adding only binary clauses may not be such a great idea so I changed it to add 3-long clauses, it exercises a different (potentially more expensive) code path :)

**Comment by msoos** [comment]
OK, pushed some more changes, it should be even faster adding clauses now :) I think I shaved off a good 20% the past 2 hours... it's 2:37am so I gotta hit the bed though. Please let me know if it helped!

**Comment by msoos** [comment]
On this note, can I license that test code you have provided as MIT? I want to include it in GIT so I can use it for performance testing and it doesn't get lost :) Please let me know. Thanks!

**Comment by mbargull** [comment]
> it's 2:37am so I gotta hit the bed though.

Hey, we're in the same timezone :) -- so I agree, heading to bed is a good idea! I'll have a look at and reply to changes/comment tomorrow when my brain is perceptive again ;).
Thanks again for all your effort!

> On this note, can I license that test code you have provided as MIT?

Yes, please consider everything from me in this repository, whether in Git itself or comments, as contributions under the project's license!

**Comment by mcg1969** [comment]
Thanks to both of you for this impressive collaboration!

**Comment by mbargull** [comment]
> (First a general note: you people are crazy good at this, where the hell did you learn to do this so professionally? I thought I was pretty OK, but damn)

Oh, it's merely some hide and seek with "try to find things like copy (-assignment) constructors in frequently executed code paths" -- child's play in comparison to creating crazy-fast SAT solvers. So I think that "pretty OK" statement is likely a "slight" understatement :laughing: 

> For 1: I don't mind it actually. We can think about it later, but it seems OK.

Nice, I'll probably try to see if I can refer polishing this part to someone more knowledgeable wrt the Python C-API then. @jakirkham, @mcg1969, anyone of you interested in picking this up and write a more polished interface that allows us to pass on clauses to `add_clauses` not only as a `List[Tuple[Int]]` but also as a flat integer array with each clause being terminated by a zero. It would be nice if this would then use https://docs.python.org/3/c-api/buffer.html while still supporting `array.array` under Python 2.

> For 2: I agree, it looks perfect!

https://github.com/msoos/cryptominisat/pull/517

----
> Thanks, I have now pushed a few commits (I made them related to #513 so you can track). It should copy less now. The performance difference for the startup remains vs pycosat but it's reduced.

Nice, I'll add some updated numbers after this post.

> Thanks for the script by the way, cool stuff, it helped -- though adding only binary clauses may not be such a great idea so I changed it to add 3-long clauses, it exercises a different (potentially more expensive) code path :)

Yes, that makes very much sense -- I was just being lazy :wink:.

----
> > On this note, can I license that test code you have provided as MIT?
> 
> Yes, please consider everything from me in this repository, whether in Git itself or comments, as contributions under the project's license!

Oh, didn't notice before that parts of the repo, i.e., helper scripts and things, are GPL'd. So "project's license" was a little imprecise and I meant, just like you enquired, MIT of course, i.e.:
Yes, please consider everything from me in this repository, whether in Git itself or comments, as contributions under the MIT license. :smile:

**Comment by mbargull** [comment]
> OK, pushed some more changes, it should be even faster adding clauses now :) I think I shaved off a good 20% the past 2 hours... it's 2:37am so I gotta hit the bed though. Please let me know if it helped!

Pretty nice! If we combine all those improvements on `master` with gh-517 and gh-518, we are just shy of having cut the `add_clauses` time in half in comparison to version `5.6.4`!

```
$ $ # current master (288a4a7) + max_var (PR #517) + use add_clause_int_tmp_cl in addClause (PR #518) + ClauseArray
$ conda create -nx --dry-run -v anaconda
Solving environment: ...working... === time_recorder total time and calls ===
TOTAL   116.922       853 conda.common.logic.Clauses.LinearBound_
TOTAL    75.254        62 conda.common.logic.Clauses._run_sat
TOTAL   189.779        12 conda.common.logic.Clauses.minimize
TOTAL    75.258        62 conda.common.logic.Clauses.sat
TOTAL     7.512        62 conda.common.logic.CryptoMiniSatSolver.invoke
TOTAL     1.690        62 conda.common.logic.CryptoMiniSatSolver.process_solution
TOTAL    32.899        62 conda.common.logic.CryptoMiniSatSolver.setup
TOTAL    24.325        62 conda.common.logic.PycoSatSolver.setup
TOTAL    46.054        62 conda.common.logic.SatSolver.run
TOTAL    21.446         2 conda.resolve.Resolve.gen_clauses
TOTAL    14.454         2 conda.resolve.Resolve.get_reduced_index
TOTAL   203.803         1 conda.resolve.Resolve.solve
```

```
$ $ # current master (288a4a7) + max_var (PR #517) + use add_clause_int_tmp_cl in addClause (PR #518) + ClauseList
$ conda create -nx --dry-run -v anaconda
Solving environment: ...working... === time_recorder total time and calls ===
TOTAL    84.291       853 conda.common.logic.Clauses.LinearBound_
TOTAL   122.308        62 conda.common.logic.Clauses._run_sat
TOTAL   205.663        12 conda.common.logic.Clauses.minimize
TOTAL   122.311        62 conda.common.logic.Clauses.sat
TOTAL     7.602        62 conda.common.logic.CryptoMiniSatSolver.invoke
TOTAL     1.718        62 conda.common.logic.CryptoMiniSatSolver.process_solution
TOTAL    64.176        62 conda.common.logic.CryptoMiniSatSolver.setup
TOTAL    39.855        62 conda.common.logic.PycoSatSolver.setup
TOTAL    77.437        62 conda.common.logic.SatSolver.run
TOTAL    21.587         2 conda.resolve.Resolve.gen_clauses
TOTAL    14.590         2 conda.resolve.Resolve.get_reduced_index
TOTAL   221.888         1 conda.resolve.Resolve.solve
```

```
TOTAL    64.176        62 conda.common.logic.CryptoMiniSatSolver.setup
TOTAL    39.855        62 conda.common.logic.PycoSatSolver.setup
```

```
TOTAL    32.899        62 conda.common.logic.CryptoMiniSatSolver.setup
TOTAL    24.325        62 conda.common.logic.PycoSatSolver.setup
```

```
$ $ # current master (288a4a7) + max_var (PR #517) + use add_clause_int_tmp_cl in addClause (PR #518)
$ python -c '...'  # script from https://github.com/msoos/cryptominisat/pull/515#issuecomment-415791578
setup CMS     array 0.180 0.180 0.181 0.182 0.183
setup CMS     list  0.322 0.346 0.358 0.359 0.370
setup Pycosat array 0.142 0.142 0.143 0.144 0.145
setup Pycosat list  0.213 0.216 0.217 0.218 0.222
solve CMS     array 0.083 0.084 0.085 0.085 0.085
solve CMS     list  0.083 0.087 0.094 0.095 0.095
solve Pycosat array 1.060 1.062 1.063 1.066 1.076
solve Pycosat list  1.057 1.059 1.062 1.067 1.068
```

**Comment by mbargull** [comment]
So, a TLDR; regarding the current state is:
We started with tests based on `5.6.4` and got the following run times:
```
$ # 5.6.4 + python-array + ClauseArray
TOTAL    62.973        62 conda.common.logic.CryptoMiniSatSolver.setup
TOTAL    24.026        62 conda.common.logic.PycoSatSolver.setup
```
```
$ # 5.6.4 + python-array + ClauseList
TOTAL   119.002        62 conda.common.logic.CryptoMiniSatSolver.setup
TOTAL    39.833        62 conda.common.logic.PycoSatSolver.setup
```
```
# 5.6.4 + python-array
setup CMS     array 0.334 0.334 0.339 0.340 0.352
setup CMS     list  0.571 0.581 0.597 0.606 0.611
setup Pycosat array 0.140 0.141 0.142 0.142 0.143
setup Pycosat list  0.207 0.216 0.217 0.218 0.224
solve CMS     array 0.094 0.094 0.095 0.095 0.099
solve CMS     list  0.083 0.090 0.094 0.095 0.096
solve Pycosat array 1.055 1.060 1.064 1.064 1.068
solve Pycosat list  1.052 1.053 1.054 1.063 1.076
```

And the same tests with `master` (288a4a7) + gh-517 + gh-518 gives us now:
```
$ $ # current master (288a4a7) + max_var (PR #517) + use add_clause_int_tmp_cl in addClause (PR #518) + ClauseArray
TOTAL    32.899        62 conda.common.logic.CryptoMiniSatSolver.setup
TOTAL    24.325        62 conda.common.logic.PycoSatSolver.setup
```
```
$ $ # current master (288a4a7) + max_var (PR #517) + use add_clause_int_tmp_cl in addClause (PR #518) + ClauseList
TOTAL    64.176        62 conda.common.logic.CryptoMiniSatSolver.setup
TOTAL    39.855        62 conda.common.logic.PycoSatSolver.setup
```
```
$ $ # current master (288a4a7) + max_var (PR #517) + use add_clause_int_tmp_cl in addClause (PR #518)
$ python -c '...'  # script from https://github.com/msoos/cryptominisat/pull/515#issuecomment-415791578
setup CMS     array 0.180 0.180 0.181 0.182 0.183
setup CMS     list  0.322 0.346 0.358 0.359 0.370
setup Pycosat array 0.142 0.142 0.143 0.144 0.145
setup Pycosat list  0.213 0.216 0.217 0.218 0.222
solve CMS     array 0.083 0.084 0.085 0.085 0.085
solve CMS     list  0.083 0.087 0.094 0.095 0.095
solve Pycosat array 1.060 1.062 1.063 1.066 1.076
solve Pycosat list  1.057 1.059 1.062 1.067 1.068
```

**Comment by msoos** [comment]
Hey :) Thanks so much, we have really made a lot of progress!

So we have merged in everything I think except for `_add_clauses_array`. Should we do that now? Should I merge it by hand or do you want to do it? Also, there is a small improvement between 288a4a7 and current master ( 380421e3e1aa2e2a635b327448f555c5c5da6716 ). So that will figure in too I guess :) It's a huge improvement I think, almost twice as fast. Thank you so much, it's been really interesting and useful.

**Comment by mbargull** [comment]
> So we have merged in everything I think except for `_add_clauses_array`. Should we do that now? Should I merge it by hand or do you want to do it?

I opened https://github.com/msoos/cryptominisat/pull/519 for this with a more cleaned up version and less jumbled Git history.

> Also, there is a small improvement between 288a4a7 and current master ( 380421e ). So that will figure in too I guess :) It's a huge improvement I think, almost twice as fast. Thank you so much, it's been really interesting and useful.

Thanks to you, too!
So this is what we get with https://github.com/msoos/cryptominisat/pull/519/commits/6ede66742594595c8b2a3a8989f122aa384dcfca:
```
setup CMS     array 0.180 0.181 0.181 0.183 0.183
setup CMS     list  0.322 0.340 0.341 0.350 0.357
setup Pycosat array 0.142 0.142 0.143 0.144 0.145
setup Pycosat list  0.214 0.214 0.216 0.216 0.222
solve CMS     array 0.084 0.085 0.086 0.089 0.090
solve CMS     list  0.084 0.087 0.094 0.096 0.096
solve Pycosat array 1.054 1.054 1.059 1.060 1.061
solve Pycosat list  1.052 1.054 1.056 1.057 1.065
```
```
$ conda create -nx --dry-run -v anaconda
Solving environment: ...working... === time_recorder total time and calls ===
TOTAL   113.101       853 conda.common.logic.Clauses.LinearBound_
TOTAL    70.288        62 conda.common.logic.Clauses._run_sat
TOTAL   181.080        12 conda.common.logic.Clauses.minimize
TOTAL    70.291        62 conda.common.logic.Clauses.sat
TOTAL     7.302        62 conda.common.logic.CryptoMiniSatSolver.invoke
TOTAL     1.667        62 conda.common.logic.CryptoMiniSatSolver.process_solution
TOTAL    29.968        62 conda.common.logic.CryptoMiniSatSolver.setup
TOTAL    22.769        62 conda.common.logic.PycoSatSolver.setup
TOTAL    42.804        62 conda.common.logic.SatSolver.run
TOTAL    19.812         2 conda.resolve.Resolve.gen_clauses
TOTAL    14.213         2 conda.resolve.Resolve.get_reduced_index
TOTAL   194.152         1 conda.resolve.Resolve.solve
```
```
TOTAL    29.968        62 conda.common.logic.CryptoMiniSatSolver.setup
TOTAL    22.769        62 conda.common.logic.PycoSatSolver.setup
```
:tada:

**Comment by mbargull** [comment]
Closing in favor of gh-519.

**Comment by msoos** [comment]
Thanks! So CMS is about 27% slower than pycosat. I say that's pretty good actually given how much more CMS does :)

**Comment by mbargull** [comment]
Yeah, I'm also quite pleased how close we've got to those `pycosat` numbers in those few days. It's been really nice working with you on this!

**Comment by msoos** [comment]
Thanks, same here :) Also, thanks for all the hard work!

**Pull request closed without merging**